File: src/sloc_counter.py
#!/usr/bin/python3

"""
Count the Source Lines of Code (SLOC) without blank lines and comments in a single source file.

CLI based tool for manual use or automation in a CI context for static code analysis.

Supported comment styles:
  Block comments (inline, multiline):       /* ... */
  Line comments  (including inline):        // ...
"""

import argparse
import re
import sys

__author__ = 'cdjaw'
__version__ = '0.2'


def parse_arguments():
    """Parse command line arguments and open input file as file object.

    :return: Parsed command line arguments
    :rtype: argparse.Namespace
    """
    parser = argparse.ArgumentParser(
        description=__doc__,
        formatter_class=argparse.RawDescriptionHelpFormatter,
    )
    parser.add_argument(
        'source_file',
        help='absolute path to the source code file to be examined',
        type=argparse.FileType('r'),
    )
    parser.add_argument(
        '-v', '--version',
        action='version',
        version='%(prog)s version ' + __version__,
    )
    return parser.parse_args()


def filter_source_lines_of_code(file_contents: str = ''):
    """Return only source lines of code without comments and blank lines.

    Limitations: The filter might undesirably strip comment-like phrases that are wrapped in a larger structure.

    :param str file_contents: Full source file contents to be filtered
    :return: Source lines of code without line endings
    :rtype: list[str]
    """
    source_lines_of_code = []
    file_contents = re.sub(r'/\*[\s\S]*?\*/', '', file_contents)                # filter block comments
    for line in file_contents.splitlines():
        line = re.sub(r'//.*$', '', line)                                       # filter line comments
        if len(line.strip()) > 0:                                               # filter blank lines
            source_lines_of_code.append(line)
    return source_lines_of_code


def main():
    """Count source lines of code in input source code file and print result to stdout.

    :return: System error code
    """
    arguments = parse_arguments()
    file_contents = arguments.source_file.read()
    source_lines_of_code = filter_source_lines_of_code(file_contents)
    print('SLOC:', len(source_lines_of_code))
    return 0


if __name__ == '__main__':
    sys.exit(main())

File: src/SlocCounter.java
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;


/**
 * *
 * A Source Line of Code (SLOC) counting tool.
 *
 * @author cdjaw
 * @version  0.1
 */
public class SlocCounter {

    public static String readFileAsString(String absolutFilePath) throws Exception {
        return new String(Files.readAllBytes(Paths.get(absolutFilePath)));
    }

    public static String removeBlockComments(String sourceFileContents) {
        return sourceFileContents.replaceAll("(?s)/\\*.*?\\*/", "");
    }

    public static boolean lineContainsSourceCode(String line) {
        String lineWithoutComments = line.replaceAll("//.*$", "");
        return lineWithoutComments.trim().length() > 0;
    }

    public static ArrayList<String> filterSourceLinesOfCode(String sourceFileContents) {
        ArrayList<String> sourceLinesOfCode;
        String[] remainingLinesOfCode;
        String remainingFileContents;

        remainingFileContents = removeBlockComments(sourceFileContents);
        remainingLinesOfCode = remainingFileContents.split("\n");
        sourceLinesOfCode = new ArrayList<>();

        for(String temp: remainingLinesOfCode) {
            if(lineContainsSourceCode(temp)) {
                sourceLinesOfCode.add(temp);
            }
        }
        return sourceLinesOfCode;
    }

    public static int countSourceLinesOfCode(String absoluteFilePath) throws Exception {
        String fileContents;
        ArrayList<String> sourceLinesOfCode;

        fileContents = readFileAsString(absoluteFilePath);
        sourceLinesOfCode = filterSourceLinesOfCode(fileContents);
        return sourceLinesOfCode.size();
    }

    public static void main(String[] args) {
        // TODO - add command line parser

        if(args.length != 1) {
            System.err.println("SlocCounter: error: requires exactly one parameter");
            System.out.println("SlocCounter: usage: java SlocCounter abs_file_path");
            System.exit(1);
        }

        try {
            System.out.println("SLOC:" + countSourceLinesOfCode(args[0]));
        } catch (Exception e) {
            System.err.println("SlocCounter: error: "+ e);
            System.exit(2);
        }
    }
}
